File: HISTORY - CODE/Comatose v1.0 (comatose3.zip) r3/C++/symbol.cpp
#include "symbol.h"

symbols::symbols(symbol sym)
	: sym_(sym)
{}

symbols::symbols(char c)
{
	from(c);
}

symbols::~symbols()
{}

symbol& symbols::sym()
{
	return sym_;
}

char symbols::to() const
{
	switch (sym_)
	{
	case symbol::C:
		return 'C';
	case symbol::P:
		return 'P';
	case symbol::BD:
		return 'X';
	case symbol::E:
		return ' ';
	case symbol::CE:
		return '*';
	case symbol::B:
		return 'B';
	case symbol::F:
		return 'F';
	default:
		return ' ';
	}
}

symbol& symbols::from(char c)
{
	switch (c)
	{
	case 'C':
		sym_ = symbol::C;
		break;
	case 'P':
		sym_ = symbol::P;
		break;
	case 'X':
		sym_ = symbol::BD;
		break;
	case ' ':
		sym_ = symbol::E;
		break;
	case '*':
		sym_ = symbol::CE;
		break;
	case 'B':
		sym_ = symbol::B;
		break;
	case 'F':
		sym_ = symbol::F;
		break;
	default:
		sym_ = symbol::E;
		break;
	}
	return sym_;
}

symbol& symbols::operator = (char c)
{
	return from(c);
}

bool symbols::operator == (const symbol sym) const
{
	return sym == sym_;
}

bool symbols::operator != (const symbol sym) const
{
	return sym != sym_;
}

arrow symbols::ar(char c)
{
	switch (c)
	{
	case 'w':
		return arrow::U;
	case 's':
		return arrow::D;
	case 'd':
		return arrow::R;
	case 'a':
		return arrow::L;
	default:
		return arrow::BAD; //WARNING
	}
}

File: bin/pyComa/symbol.py
from getch import getch

class Enum(object):
    def __init__(self, val_out):
        self.__values__ = []
        self.__out__ = []
        self.__to__ = {}
        self.__from__ = {}
        for val, out in val_out:
            self.__values__.append(val)
            self.__out__.append(out)
            self.__to__.update({val:out})
            self.__from__.update({out:val})
    def to(self, val):
        return self.__to__.get(val)
    def fromm(self, out):
        return self.__from__.get(out)
    def get(self, val):
        if val in self.__values__:
            return val
    def __getattr__(self, val):
        if not val in self.__values__:
            raise AttributeError('value {} does not exist'.format(val))
        return self.get(val)

symbol = Enum([
    ('C', 'C'),
    ('P', 'P'),
    ('BD', 'X'),
    ('E', ' '),
    ('CE', '*'),
    ('B', 'B'),
    ('F', 'F'),
])

arrow = Enum([
    ('U', 'w'),
    ('D', 's'),
    ('R', 'd'),
    ('L', 'a'),
    ('BAD', None) #WARNING
])

class symbols(object):
    def __init__(self, sym):
        self.set(sym)
    def sym(self, symm=None):
        if symbol.get(symm):
            self.sym_ = symm
        return self.sym_
    def to(self):
        return symbol.to(self.sym_)
    def fromm(self, c):
        self.sym_ = symbol.fromm(c)
        return self.sym_
    def set(self, sym):
        if isinstance(sym, symbols):
            self.sym_ = sym.sym_
        elif symbol.get(sym):
            self.sym_ = sym
        else:
            self.fromm(sym)
    @classmethod
    def ar(cls, c):
        return arrow.fromm(c)
    def __eq__(self, other):
        if isinstance(other, symbols):
            return self.sym_ == other.sym_
        else:
            return self.sym_ == other
    def __ne__(self, other):
        return not self == other
'''
    symbol& symbols::operator = (char c)
    {
        return fromm(c);
    }
'''
